packages/ecr-credential-provider/ecr-credential-provider.spec

```diff
@@ -0,0 +1,56 @@
+%global _missing_build_ids_terminate_build 0
```

Comment:
This might not be necessary with a more typical `go build` invocation, like other specs, where `-ldflags "${GOLDFLAGS}"` is set.

Reply:
I was getting an error about the binary file not having a build ID. I would need to find the issue link again, but it appears to be a known issue with RPM packaging of Go binaries. This was the accepted workaround to prevent the error.

That said, there were a few things I was working through at the time, so I will try removing this and see if it is no longer an issue.

Reply:
Just removing the top statement:

```
  #19 12.25 RPM build errors:
  #19 12.25     Missing build-id in /home/builder/rpmbuild/BUILDROOT/bottlerocket-x86_64-aws-ecr-credential-provider-1.24.1-1.aarch64/x86_64-bottlerocket-linux-gnu/sys-root/usr/share/factory/var/lib/kubelet/plugins/ecr-credential-provider
  #19 12.25     Generating build-id links failed
```

Adding `GOLDFLAGS`:

```
  #19 8.461 loadinternal: cannot find runtime/cgo
  #19 8.461 /usr/libexec/go/pkg/tool/linux_arm64/link: running x86_64-bottlerocket-linux-gnu-gcc failed: exit status 1
  #19 8.461 /usr/lib/gcc/x86_64-bottlerocket-linux-gnu/10/../../../../x86_64-bottlerocket-linux-gnu/bin/ld: cannot find Scrt1.o: No such file or directory
  #19 8.461 /usr/lib/gcc/x86_64-bottlerocket-linux-gnu/10/../../../../x86_64-bottlerocket-linux-gnu/bin/ld: cannot find crti.o: No such file or directory
  #19 8.461 collect2: error: ld returned 1 exit status
```

Removing `CGO_ENABLED=0`:

```
  #19 0.605 + go build -mod=vendor -trimpath -a -ldflags '-compressdwarf=false -linkmode=external' -o=ecr-credential-provider cmd/ecr-credential-provider/main.go cmd/ecr-credential-provider/main_test.go cmd/ecr-credential-provider/plugin.go cmd/ecr-credential-provider/plugin_test.go
  #19 2.553 # runtime/cgo
  #19 2.553 _cgo_export.c:3:10: fatal error: stdlib.h: No such file or directory
  #19 2.553     3 | #include <stdlib.h>
  #19 2.553       |          ^~~~~~~~~~
  #19 2.553 compilation terminated.
```

So looks like it may need to stay as it is. Unless there's another part I'm missing here.

Reply:
Hmm, I will need to look at this - it indicates that something isn't being built with our expected toolchain. Those linker errors or missing headers shouldn't happen when the build flags are correct.

Reply:
Hmm, I will need to look at this - it indicates that something isn't being built with our expected toolchain. Those linker errors or missing headers shouldn't happen when the build flags are correct.

Reply:
Ah, of course - the spec needs a dependency on glibc for cgo and the normal linker flags to work.

```
diff --git a/packages/ecr-credential-provider/Cargo.toml b/packages/ecr-credential-provider/Cargo.toml
index 35e255f6..eb488b7f 100644
--- a/packages/ecr-credential-provider/Cargo.toml
+++ b/packages/ecr-credential-provider/Cargo.toml
@@ -15,3 +15,6 @@ releases-url = "https://github.com/kubernetes/cloud-provider-aws/releases"
 url = "https://github.com/kubernetes/cloud-provider-aws/archive/b9c8d82db9edc3be45f82f8d5320d006c62aa50c/v1.24.1.tar.gz"
 sha512 = "d82ab5f9b05db8bcaf864faf87597d0f17cb9a4e52b0d192611725bf7e1c1512fe604b4a0aea0761058eca8ed6c1d0a96f6575d745b6082b9ad78c1290504990"
 bundle-modules = [ "go" ]
+
+[build-dependencies]
+glibc = { path = "../glibc" }
diff --git a/packages/ecr-credential-provider/ecr-credential-provider.spec b/packages/ecr-credential-provider/ecr-credential-provider.spec
index 235aeb93..c7d5be91 100644
--- a/packages/ecr-credential-provider/ecr-credential-provider.spec
+++ b/packages/ecr-credential-provider/ecr-credential-provider.spec
@@ -1,6 +1,3 @@
-%global _missing_build_ids_terminate_build 0
-%define debug_package %{nil}
-
 %global goproject github.com/kubernetes
 %global gorepo cloud-provider-aws
 %global goimport %{goproject}/%{gorepo}
@@ -8,6 +5,8 @@
 %global gover 1.24.1
 %global rpmver %{gover}
 
+%global _dwz_low_mem_die_limit 0
+
 %global gitrev b9c8d82db9edc3be45f82f8d5320d006c62aa50c
 %global shortrev %(c=%{gitrev}; echo ${c:0:7})
 
@@ -23,6 +22,8 @@ Source1: bundled-v%{gover}.tar.gz
 Source100: root-.aws.mount
 Source1000: clarify.toml
 
+BuildRequires: %{_cross_os}glibc-devel
+
 %description
 %{summary}.
 
@@ -32,9 +33,7 @@ Source1000: clarify.toml
 
 %build
 %set_cross_go_flags
-
-# Set CGO_ENABLED=0 to statically link ecr-credential-provider.
-CGO_ENABLED=0 go build ${GOFLAGS} -trimpath -a -ldflags "-s" -o=ecr-credential-provider cmd/ecr-credential-provider/*.go
+go build -buildmode=pie -ldflags="${GOLDFLAGS}" -o=ecr-credential-provider cmd/ecr-credential-provider/*.go
 
 %install
 install -d %{buildroot}%{_cross_libexecdir}/kubernetes/kubelet/plugins
diff --git a/variants/Cargo.lock b/variants/Cargo.lock
index 1c64e2a6..b2993a37 100644
--- a/variants/Cargo.lock
+++ b/variants/Cargo.lock
@@ -285,6 +285,9 @@ dependencies = [
 [[package]]
 name = "ecr-credential-provider"
 version = "0.1.0"
+dependencies = [
+ "glibc",
+]
 
 [[package]]
 name = "ecs-agent"
```

You should also drop `-trimpath -a -ldflags "-s"` so that the debuginfo is added. (It will get removed later as part of the standard debuginfo stripping that happens in an rpm build.)

Reply:
That works! Thanks for digging in to that.